### PR TITLE
MULE-13138: Decouple runtime and spring module from spring libraries

### PR DESCRIPTION
**Pull request by elrodro83**
* Remove asm from Distro

Follow up tasks:
* https://www.mulesoft.org/jira/browse/MULE-13203
* https://www.mulesoft.org/jira/browse/MULE-13204
* https://www.mulesoft.org/jira/browse/MULE-13248

### CODE REVIEW

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/BeanDefinitionFactory.java [hunk 1]
```diff
@@ -149,23 +150,24 @@ public void resolveEagerCreationObjects(ComponentModel parentComponentModel, Com
...
                                                     BeanDefinitionRegistry registry,
                                                     BiConsumer<ComponentModel, BeanDefinitionRegistry> componentModelPostProcessor,
                                                     BiFunction<Element, BeanDefinition, Either<BeanDefinition, BeanReference>> oldParsingMechanism) {
     List<ComponentModel> innerComponents = componentModel.getInnerComponents();
     if (!innerComponents.isEmpty()) {
       for (ComponentModel innerComponent : innerComponents) {
         if (innerComponent.isEnabled()) {
-          resolveComponentRecursively(componentModel, innerComponent, registry, componentModelPostProcessor,
-                                      oldParsingMechanism);
+          resolveComponentRecursively(componentModel, (SpringComponentModel) innerComponent, registry,
```
Comment (pablokraan) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/BeanDefinitionFactory.java]: Should not check that the component is of the expected class before casting it? (There are many places where this occurs)

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/ComponentConfigurationBuilder.java [hunk 2]
```diff
@@ -79,13 +81,14 @@ public void processConfiguration() {
     return componentModel.getInnerComponents().stream().filter(child -> child.isEnabled()).map(cdm -> {
       // When it comes from old model it does not have the type set
       Class<?> beanDefinitionType = cdm.getType();
+      final SpringComponentModel csdm = (SpringComponentModel) cdm;
```
Comment (pablokraan) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/ComponentConfigurationBuilder.java]: Can you use a more meaningful names for this variable?

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/ObjectFactoryClassRepository.java [hunk 3]
```diff
@@ -126,4 +140,78 @@ public void destroy() {
...
 
+  /**
+   * Classloader implementation that, given a set of classloaders, will first search for a resource/class in the first one. If it
+   * is not found, it will try the nextx and so on until the resource/class is found or all classloaders have been tried.
+   * <p>
+   * For {@link #getResources(String)}, all the classloders will be queried to get the union of all found resources.
+   * 
+   * @since 1.0
+   */
+  private static class CompositeClassLoader extends ClassLoader {
```
Comment (pablokraan) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/ObjectFactoryClassRepository.java]: This class needs unit tests

File: modules/extensions-spring-support/src/main/resources/META-INF/mule-module.properties [hunk 4]
```diff
@@ -1,14 +1,22 @@
 module.name=extensions-spring-support
 
+# TODO MULE-13204 Remove dom4j and jaxen packages
 artifact.export.classPackages=org.mule.runtime.module.extension.internal.capability.xml.schema,\
                               org.mule.runtime.module.extension.internal.capability.xml.schema.builder,\
                               org.mule.runtime.module.extension.internal.capability.xml.schema.model,\
-                              org.mule.runtime.module.extension.internal.runtime.execution
+                              org.mule.runtime.module.extension.internal.runtime.execution,\
+                              org.dom4j.rule,\
```
Comment (elrodro83) [inline modules/extensions-spring-support/src/main/resources/META-INF/mule-module.properties]: There's a TODO for this: https://www.mulesoft.org/jira/browse/MULE-13204

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/BeanDefinitionFactory.java (hunk 1 above)
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/internal/dsl/spring/BeanDefinitionFactory.java]: Made the superclass abstract, and SpringComponentModel is the only subclass. It should go away when the spring stuff is not needed here anymore

**Pull request merged**